allensdk/brain_observatory/ecephys/lfp_subsampling/subsampling.py

```diff
@@ -0,0 +1,245 @@
+# POSSIBILITY OF SUCH DAMAGE.
+#
+import logging
+
+import numpy as np
+from scipy.signal import decimate, butter, filtfilt
+
+
+logger = logging.getLogger(__name__)
+
+
+def select_channels(total_channels,
+                    surface_channel,
+                    surface_padding,
+                    start_channel_offset,
+                    channel_stride,
+                    channel_order,
+                    noisy_channels=np.array([]),
+                    remove_noisy_channels=False,
+                    reference_channels=np.array([]),
+                    remove_references=False):
+    """
+    Selects a subset of channels for spatial downsampling
+
+    Parameters:
+    ----------
+
+    total_channels : int
+        Number of channels in the original data file
+    surface_channel : int
+        Index of channel at brain surface
+    surface_padding : int
+        Number of channels above surface to save
+    start_channel_offset : int
+        First channel to save
+    channel_stride : int
+        Number of channels to skip in output
+    channel_order : np.ndarray
+        Actual order of LFP channels (needed to account for the bug in NPX extraction)
+    noisy_channels : numpy.ndarray
+        Array indicating noisy channels
+    remove_noisy_channels : bool
+        Flag to remove noisy channels
+    reference_channels : numpy.ndarray
+        Array indicating refence channels
+    remove_references : bool
+        Flag to remove reference channels
+
+    Returns:
+    --------
+    selected_channels : numpy.ndarray
+        Indices of channels to select (relative to non-remapped data)
+    actual_channel_numbers : numpy.ndarray
+        Actual probe channels in subsampled data
+
+    """
+    assert surface_channel <= total_channels
+
+    max_channel = np.min([total_channels, surface_channel + surface_padding])
+
+    selected_channels = channel_order[start_channel_offset:max_channel:channel_stride]
+
+    actual_channel_numbers = np.arange(total_channels)
+    actual_channel_numbers = actual_channel_numbers[start_channel_offset:max_channel:channel_stride]
+
+    if remove_references:
+        mask = np.isin(actual_channel_numbers, reference_channels, assume_unique=True, invert=True)
+        actual_channel_numbers = actual_channel_numbers[mask]
+        selected_channels = selected_channels[mask]
+
+    logger.info("Before:")
+    logger.info(actual_channel_numbers)
+
+    logger.info("To remove:")
+    logger.info(noisy_channels)
+
+    if remove_noisy_channels:
+        mask = np.isin(actual_channel_numbers, noisy_channels, assume_unique=True, invert=True)
+        logger.info(np.where(mask)[0])
+        actual_channel_numbers = actual_channel_numbers[mask]
+        selected_channels = selected_channels[mask]
+
+    logger.info("After:")
+    logger.info(actual_channel_numbers)
+
+    # stop
+
+    return selected_channels, actual_channel_numbers
+
+
+def subsample_timestamps(timestamps, subsampling_factor):
```

Comment:
Any reason not to inline this?

Reply:
Someone needed a helper function so I'm inclined to leave it in. Plus having it doesn't require users to assume the data structure of timestamps, which could be helpful if it changes in the future.

Reply:
OK. My preference is definitely to inline functions that are 1 line long and called only once, but I can see why you would want to leave this open to configuration.